core/silkworm/consensus/clique/clique_snapshot.cpp

```diff
@@ -0,0 +1,248 @@
+    auto hash{header.hash()};
+    auto tmp_recents{recents_}; // We only modify snapshot after checks are done.
+    // Delete the oldest signer from the recent list to allow it signing again
+    if (recents_.size() > signers_.size() / 2) {
+        tmp_recents.pop_back();
+    }
+
+    if (std::find(signers_.begin(), signers_.end(), signer) == signers_.end()) {
+        return ValidationResult::kUnauthorizedSigner;
+    }
+
+    // Remove any votes on checkpoint blocks
+    if (header.number % config.epoch == 0) {
+        tallies_.clear();
+    }
+
+    if (std::find(tmp_recents.begin(), tmp_recents.end(), signer) != tmp_recents.end()) {
+        return ValidationResult::kRecentlySigned;
+    }
+
+    // We check what the vote is
+    bool authorize;
+    if (header.nonce == kNonceAuthorize) {
+        authorize = true;
+    } else if (header.nonce == kNonceUnauthorize) {
+        authorize = false;
+    } else {
+        return ValidationResult::kInvalidVote;
+    }
+    
+    tmp_recents.push_front(signer);
+    recents_ = tmp_recents;
+
+    if (!header.beneficiary) {
+        update(header.number, hash);
+        return ValidationResult::kOk;
+    }
+
+    // Uncast votes from signers
+    decrement_vote(header.beneficiary, signer);
+    // do casting
+    increment_vote(header.beneficiary, signer, authorize);
+    // If the vote passed, update the list of signers
+    auto current_tally{tallies_[header.beneficiary]};
+    if (current_tally.votes > signers_.size() / 2) {
+        if (current_tally.authorize) {
+            signers_.push_back(header.beneficiary);
+        } else {
+            signers_.erase(std::remove(signers_.begin(), signers_.end(), header.beneficiary), signers_.end());
+            // Clean up recents
+            recents_.pop_back();
+            // Update Tallies
+            clear_votes(header.beneficiary);
+        }
+        if (tallies_.count(header.beneficiary)) {
+            tallies_.erase(header.beneficiary);
+        }
+    }
+
+    update(header.number, hash);
+    // Success
+    return ValidationResult::kOk;
+}
+
+//! \brief Verify seal for header
+//! \param header: header to verify.
+ValidationResult CliqueSnapshot::verify_seal(const BlockHeader& header, const evmc::address& signer) {
+    // Check difficulty
+    auto authority{is_authority(header.number, signer)};
+    if (authority && header.difficulty != kDiffInTurn && header.difficulty != kDiffNoTurn) {
+        return ValidationResult::kIntrinsicGas;
+    }
+
+    return ValidationResult::kOk;
+}
+
+//! \brief Checks for authority
+//! \param block_number: Block to check.
+//! \param address: Address to check.
+//! \return if a signer at a given block height is in charge or not.
+bool CliqueSnapshot::is_authority(uint64_t block_number, evmc::address address) const noexcept {
+    uint64_t offset{0};
+	while (offset < signers_.size() && signers_[offset] != address) {
+		++offset;
+	}
+	return (block_number % signers_.size()) == offset;
+}
```

Comment:
This is wrong. You increase complexity to O(n) while complexity is O(1).
This time I won't put the suggestion here as I wish you reason a bit on that